### PR TITLE
Adds Planning and Parsing Support for Drop Table, Drop View and Alter View

### PR DESCRIPTION
**Pull request by GuptaManan100**
Adds Planning and Parsing Support for `Drop Table`, `Drop View` and `Alter View`

### CODE REVIEW

File: go/vt/sqlparser/ast.go [hunk 1]
```diff
@@ -438,11 +445,13 @@ func (*CreateTable) iStatement()       {}
 func (*CreateView) iStatement()        {}
 func (*LockTables) iStatement()        {}
 func (*UnlockTables) iStatement()      {}
+func (*DropTable) iStatement()         {}
 
 func (*DDL) iDDLStatement()         {}
 func (*CreateIndex) iDDLStatement() {}
 func (*CreateView) iDDLStatement()  {}
 func (*CreateTable) iDDLStatement() {}
+func (*DropTable) iDDLStatement()   {}
```
Comment (shlomi-noach) [inline go/vt/sqlparser/ast.go:454]: What do `iStatement` and `iDDLStatement` mean?

**Review by shlomi-noach — APPROVED** [review]

File: go/vt/sqlparser/ast.go (hunk 1 above)
Comment (GuptaManan100) [reply go/vt/sqlparser/ast.go:454]: `iStatement` is an empty function that is needed to implement the Statement interface. Without this all the SQLNodes would also implement statement interface which we dont want. `iDDLStatement` is the DDL counterpart of `iStatement`. This just makes it super clear all the statements that implement the DDLStatement interface and no other struct implements them by accident.

Comment (GuptaManan100) [reply go/vt/sqlparser/ast.go:454]: Check these out https://github.com/vitessio/vitess/blob/777a19dbe6895f8db7de0c7cccfeec279cb0ec66/go/vt/sqlparser/ast.go#L42 and https://github.com/vitessio/vitess/blob/777a19dbe6895f8db7de0c7cccfeec279cb0ec66/go/vt/sqlparser/ast.go#L60

Comment (shlomi-noach) [reply go/vt/sqlparser/ast.go:454]: Got it, thanks for explaining!

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vtgate/executor_test.go [hunk 2]
```diff
@@ -1128,7 +1128,6 @@ func TestExecutorDDL(t *testing.T) {
 		"alter table t1 add primary key id",
 		"rename table t1 to t2",
 		"truncate table t2",
-		"drop table t2",
```
[the remark this answers is not in the document]
Comment (GuptaManan100) [reply go/vt/vtgate/executor_test.go:1131]: Yeah the reason for removing this is that the error thrown from this statement is different from earlier expected. Earlier we only used to check that the keyspace should exist but now in drop statements, we check that the table should exist. That is why I had to remove it from the test case.

File: go/vt/vtgate/planbuilder/ddl.go [hunk 3]
```diff
@@ -81,13 +81,52 @@ func buildDDLPlans(sql string, ddlStatement sqlparser.DDLStatement, vschema Cont
...
+				})
+			}
+			return true
+		}, nil)
+
+	case *sqlparser.AlterView:
+		// For Alter View, we require that the view exist and the select query can be satisfied within the keyspace itself
+		// We should remove the keyspace name from the table name, as the database name in MySQL might be different than the keyspace name
+		table, _, _, _, destination, err = vschema.FindTableOrVindex(ddl.ViewName)
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/ddl.go:106]: done in the last commit

File: go/vt/vtgate/executor_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (harshit-gangal) [reply go/vt/vtgate/executor_test.go:1131]: can you add those tests to executor tests.

File: go/vt/vtgate/planbuilder/ddl.go [hunk 4]
```diff
@@ -132,3 +114,106 @@ func buildDDLPlans(sql string, ddlStatement sqlparser.DDLStatement, vschema Cont
...
+	var destination key.Destination
+	var keyspace *vindexes.Keyspace
+	for i, tab := range ddlStatement.GetFromTables() {
+		table, _, _, _, destinationTab, err := vschema.FindTableOrVindex(tab)
+		if err != nil {
+			return nil, nil, err
+		}
+		if table == nil {
+			return nil, nil, vterrors.Errorf(vtrpc.Code_INVALID_ARGUMENT, "table does not exists: %s", tab.Name.String())
+		}
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/ddl.go:205]: I think we should not assume the the table will be present in vschema.
Also, for unsharded keyspace there is no need to provide tables in the vschema.

**Review by harshit-gangal — CHANGES_REQUESTED** [review]

File: go/vt/vtgate/planbuilder/ddl.go (hunk 4 above)
Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/ddl.go:205]: For sharded keyspace we want the table to be present in the vschema. For unsharded keyspaces it does not require that the table be present in the vschema.  There is a check for that in the implementation of the function. I have also added a test for the same.

Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/ddl.go:205]: As per discussion, will make the necessary changes

Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/ddl.go]: done in the last commit

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**